toolchain/sem_ir/formatter.cpp

```diff
@@ -78,9 +90,103 @@ class FormatterImpl {
     out_ << "\n";
   }
 
+  // Write buffered output to the given stream.
+  auto Write(llvm::raw_ostream& out) -> void {
+    FlushChunk();
+    for (const auto& chunk : output_chunks_) {
+      if (chunk.include_in_output) {
+        out << chunk.chunk;
+      }
+    }
+  }
+
  private:
   enum class AddSpace : bool { Before, After };
 
+  // A chunk of the buffered output. Chunks of the output, such as constant
+  // values, are buffered until we reach the end of formatting so that we can
+  // decide whether to include them based on whether they are referenced.
+  struct OutputChunk {
+    // Whether this chunk is known to be included in the output.
+    bool include_in_output;
+    // The textual contents of this chunk.
+    std::string chunk = std::string();
+    // Chunks that should be included in the output if this one is.
+    llvm::SmallVector<size_t> dependencies = {};
+  };
+
+  // A scope in which output should be buffered because we don't yet know
+  // whether to include it in the final formatted SemIR.
+  struct TentativeOutputScope {
+    explicit TentativeOutputScope(FormatterImpl& f) : formatter(f) {
+      if (formatter.after_open_brace_) {
+        formatter.out_ << '\n';
+        formatter.after_open_brace_ = false;
+      }
+      index = formatter.AddChunk(false);
+    }
+    ~TentativeOutputScope() {
+      auto next_index = formatter.AddChunk(true);
+      CARBON_CHECK(next_index == index + 1, "Nested TentativeOutputScope");
+    }
+    FormatterImpl& formatter;
+    size_t index;
+  };
+
+  // Flushes the buffered output to the current chunk.
+  auto FlushChunk() -> void {
+    CARBON_CHECK(output_chunks_.back().chunk.empty());
+    output_chunks_.back().chunk = std::move(buffer_);
+    buffer_.clear();
+  }
+
+  // Adds a new chunk to the output. Does not flush existing output, so should
+  // only be called if there is no buffered output.
+  auto AddChunkNoFlush(bool include_in_output) -> size_t {
+    CARBON_CHECK(buffer_.empty());
+    output_chunks_.push_back({include_in_output});
+    return output_chunks_.size() - 1;
+  }
+
+  // Flushes the current chunk and add a new chunk to the output.
+  auto AddChunk(bool include_in_output) -> size_t {
+    FlushChunk();
+    return AddChunkNoFlush(include_in_output);
+  }
+
+  // Marks the given chunk as being included in the output if the current chunk
+  // is.
+  auto IncludeChunkInOutput(size_t chunk) -> void {
+    if (chunk == output_chunks_.size() - 1) {
+      return;
+    }
+
+    if (!output_chunks_.back().include_in_output) {
+      output_chunks_.back().dependencies.push_back(chunk);
```

Comment:
Small suggestion, trying to use the variable name for self-documenting code.

```suggestion
    if (auto& parent = output_chunks_.back(); !parent.include_in_output) {
      parent.dependencies.push_back(chunk);
```

Reply:
To match the function comment, maybe:
```suggestion
    if (auto& current_chunk = !output_chunks_.back();
        current_chunk.include_in_output) {
     current_chunk.dependencies.push_back(chunk);
```